controllers/tlspolicy_status_updater.go

```diff
@@ -0,0 +1,240 @@
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/ptr"
+	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/kuadrant"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+)
+
+type TLSPolicyStatusUpdaterReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewTLSPolicyStatusUpdaterReconciler(client *dynamic.DynamicClient) *TLSPolicyStatusUpdaterReconciler {
+	return &TLSPolicyStatusUpdaterReconciler{Client: client}
+}
+
+func (t *TLSPolicyStatusUpdaterReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.TLSPolicyGroupKind, EventType: ptr.To(controller.CreateEvent)},
+			{Kind: &kuadrantv1alpha1.TLSPolicyGroupKind, EventType: ptr.To(controller.UpdateEvent)},
+			{Kind: &CertManagerCertificateKind},
+			{Kind: &CertManagerIssuerKind},
+			{Kind: &CertManagerClusterIssuerKind},
+		},
+		ReconcileFunc: t.UpdateStatus,
+	}
+}
+
+func (t *TLSPolicyStatusUpdaterReconciler) UpdateStatus(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, s *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("TLSPolicyStatusUpdaterReconciler").WithName("Reconcile")
+
+	policies := lo.FilterMap(topology.Policies().Items(), func(item machinery.Policy, index int) (*kuadrantv1alpha1.TLSPolicy, bool) {
+		p, ok := item.(*kuadrantv1alpha1.TLSPolicy)
+		return p, ok
+	})
+
+	store, ok := s.Load(TLSPolicyAcceptedKey)
+	if !ok {
+		logger.Error(errors.New("TLSPolicyAcceptedKey not found, skipping update of tls policy statuses"), "sync map error")
+		return nil
+	}
+
+	isPolicyValidErrorMap := store.(map[string]error)
+
+	for _, policy := range policies {
+		if policy.DeletionTimestamp != nil {
+			logger.V(1).Info("tls policy is marked for deletion, skipping", "name", policy.Name, "namespace", policy.Namespace)
+			continue
+		}
+
+		newStatus := &kuadrantv1alpha1.TLSPolicyStatus{
+			// Copy initial conditions. Otherwise, status will always be updated
+			Conditions:         slices.Clone(policy.Status.Conditions),
+			ObservedGeneration: policy.Status.ObservedGeneration,
+		}
+
+		err := isPolicyValidErrorMap[policy.GetLocator()]
+		meta.SetStatusCondition(&newStatus.Conditions, *kuadrant.AcceptedCondition(policy, err))
+
+		// Do not set enforced condition if Accepted condition is false
+		if meta.IsStatusConditionFalse(newStatus.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)) {
+			meta.RemoveStatusCondition(&newStatus.Conditions, string(kuadrant.PolicyConditionEnforced))
+		} else {
+			enforcedCond := t.enforcedCondition(ctx, policy, topology)
+			meta.SetStatusCondition(&newStatus.Conditions, *enforcedCond)
+		}
+
+		// Nothing to do
+		equalStatus := equality.Semantic.DeepEqual(newStatus, policy.Status)
```

Comment:
We need to be careful to never allow unexported fields in the [`TLSPolicyStatus`](https://github.com/Kuadrant/kuadrant-operator/blob/b8f0bf6b9a4a13bb0defcad29b553ab88cd810ec/api/v1alpha1/tlspolicy_types.go#L117) type, otherwise this line will start panicking.